crates/red_knot_python_semantic/src/semantic_index/use_def/bitset.rs

```diff
@@ -0,0 +1,402 @@
+    }
+
+    /// Return an iterator over the values (in ascending order) in this [`BitSet`].
+    pub(super) fn iter(&self) -> BitSetIterator<'_, B> {
+        match self {
+            Self::Inline(blocks) => BitSetIterator::Inline {
+                blocks,
+                cur_block_index: 0,
+                cur_block: blocks[0],
+            },
+            Self::Heap(set) => BitSetIterator::Heap(set.iter()),
+        }
+    }
+}
+
+/// Iterator over values in a [`BitSet`].
+#[derive(Debug)]
+pub(super) enum BitSetIterator<'a, const B: usize> {
+    Inline {
+        /// The blocks we are iterating over.
+        blocks: &'a [u64; B],
+
+        /// The index of the block we are currently iterating through.
+        cur_block_index: usize,
+
+        /// The block we are currently iterating through (and zeroing as we go.)
+        cur_block: u64,
+    },
+    Heap(btree_set::Iter<'a, u32>),
+}
+
+impl<const B: usize> Iterator for BitSetIterator<'_, B> {
+    type Item = u32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Inline {
+                blocks,
+                cur_block_index,
+                cur_block,
+            } => {
+                while *cur_block == 0 {
+                    if *cur_block_index == B - 1 {
+                        return None;
+                    }
+                    *cur_block_index += 1;
+                    *cur_block = blocks[*cur_block_index];
+                }
+                let value = cur_block.trailing_zeros();
+                // reset the lowest set bit
+                *cur_block &= cur_block.wrapping_sub(1);
+                // SAFETY: `value` cannot be more than 64, `cur_block_index` cannot be more than
+                // `B - 1`, and we check above that `B * 64 < u32::MAX`. So both `64 *
+                // cur_block_index` and the final value here must fit in u32.
+                #[allow(clippy::cast_possible_truncation)]
+                Some(value + (64 * *cur_block_index) as u32)
+            }
+            Self::Heap(set_iter) => set_iter.next().copied(),
+        }
+    }
+}
+
+impl<const B: usize> std::iter::FusedIterator for BitSetIterator<'_, B> {}
+
+/// Array of [`BitSet<B>`]. Up to N stored inline, more than that overflows to a Vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum BitSetArray<const B: usize, const N: usize> {
+    Inline {
+        /// Array of N [`BitSet`].
+        array: [BitSet<B>; N],
+
+        /// How many of the bitsets are used?
+        size: usize,
+    },
+
+    Heap(Vec<BitSet<B>>),
+}
+
+impl<const B: usize, const N: usize> Default for BitSetArray<B, N> {
+    fn default() -> Self {
+        Self::Inline {
+            array: std::array::from_fn(|_| BitSet::default()),
+            size: 0,
+        }
+    }
+}
+
+impl<const B: usize, const N: usize> BitSetArray<B, N> {
+    /// Create a [`BitSetArray`] of `size` empty [`BitSet`]s.
+    pub(super) fn of_size(size: usize) -> Self {
```

Comment:
I think that matches what in Rust commonly is named `with_capacity`

```suggestion
    pub(super) fn with_capacity(size: usize) -> Self {
```

Reply:
I think it's different; the "capacity" here is already fixed by the const generic parameter. This constructor is actually creating a non-empty array of `size` user-visible empty bitset values. The name `with_capacity` would wrongly suggest that it's creating an empty bitset array with a capacity to hold a certain number of bitsets without resizing.